dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,679 @@
-	// of this class.
-	Definition *DeviceClassDefinition `json:"definition,omitempty"`
+// Constraint must have one and only one field set.
+type Constraint struct {
+	// The match criteria must be satisfied before a set of devices will be used
+	// together.
+	//
+	// +optional
+	Match *MatchModel `json:"match,omitempty"`
 }
 
-// DeviceClassDefinition specifies the subset of devices to consider part of
-// this class, along with device-specific configuration information for those
-// devices.
-type DeviceClassDefinition struct {
-	// Driver specifies the driver that should handle this class of devices.
-	// When a DeviceClaim uses this class, only devices published by the
-	// specified driver will be considered.
-	// +required
-	Driver string `json:driver,omitempty`
+// ConfigurationParameters must have one and only one field set.
+type ConfigurationParameters struct {
+	Vendor *VendorConfigurationParameters `json:"vendor,omitempty" protobuf:"bytes,1,opt,name=vendor"`
+}
 
-	// Constraints is a CEL expression that operates on device attributes,
-	// and must evaluate to true for a device to be considered. It will be
-	// ANDed with any Constraints field in the DeviceClaim using this class.
+// VendorConfigurationParameters contains configuration parameters for a driver.
+type VendorConfigurationParameters struct {
+	// DriverName is used to determine which kubelet plugin needs
+	// to be passed these configuration parameters.
+	//
+	// An admission webhook provided by the vendor could use this
+	// to decide whether it needs to validate them.
+	DriverName string `json:"driverName,omitempty" protobuf:"bytes,1,opt,name=driverName"`
+
+	// Parameters can contain arbitrary data. It is the
+	// responsibility of the vendor to handle validation and
+	// versioning.
+	Parameters runtime.RawExtension `json:"parameters,omitempty" protobuf:"bytes,2,opt,name=parameters"`
+}
+
+// Requirement must have one and only one field set.
+type Requirement struct {
+	// Device describes a filter based on device attributes.
+	// This covers "qualitative" aspects of a device.
 	//
 	// +optional
-	Constraints *string `json:"constraints,omitempty"`
+	Device *DeviceFilter `json:"device,omitempty"`
 
-	// DeviceConfigs contains references to arbitrary vendor device configuration
-	// objects that will be attached to the device allocation.
+	// Resource describes how much of a consumable resource provided by a
+	// device are needed.
+	//
+	// FUTURE EXTENSION, not planned for 1.31!
+	Resource *ResourceRequirement `json:"resource,omitempty"`
+}
+
+type DeviceFilter struct {
+	// DriverName excludes any device not provided by this driver.
+	// Filters that are meant to match devices provided by different
+	// drivers need to leave this field blank.
 	//
 	// +optional
-	// +listType=atomic
-	Configs []DeviceClassConfigReference `json:"configs,omitempty"`
+	DriverName *string `json:"driverName,omitempty" protobuf:"bytes,1,opt,name=driverName"`
```

Comment:
This field is a convenience and/or optimization. It allows writing:

```yaml
requests:
- driverName: dra.example.com
```

This then selects any device from that driver without having to write CEL (convenience) and without having to evaluate it (optimization).

We can drop it. The corresponding selector is:

```yaml
requests:
- device:
    selector: device.driverName == "dra.example.com"
```

This "device.driverName" is not an attribute because then vendors would have to remember to set it correctly in their device attributes, or Kubernetes would need to mess with those.

Instead, this gets added to the CEL environment based on the `ResourcePool.DriverName`.

Reply:
I think I requested this "convenience" at some point over the last week, but I'd be fine dropping it if that is the concensus.

Reply:
I removed it.

Reply:
The other purpose of `DriverName` was as a fallback suffix:
```yaml
requests:
- device:
    driverName: dra.example.com
    selector: device.attributes["memory" /* = memory.dra.example.com */].isGreaterThan(quantity("1Gi"))
```